xblock/fields.py

```diff
@@ -838,31 +842,27 @@ def from_json(self, value):
         """
         Parse the date from an ISO-formatted date string, or None.
         """
-        if isinstance(value, basestring):
+        if value is None:
+            return None
 
+        if isinstance(value, basestring):
             # Parser interprets empty string as now by default
             if value == "":
                 return None
-
             try:
-                parsed_date = dateutil.parser.parse(value)
+                value = dateutil.parser.parse(value)
             except (TypeError, ValueError):
                 raise ValueError("Could not parse {} as a date".format(value))
 
-            if parsed_date.tzinfo is not None:  # pylint: disable=maybe-no-member
-                parsed_date.astimezone(pytz.utc)  # pylint: disable=maybe-no-member
-            else:
-                parsed_date = parsed_date.replace(tzinfo=pytz.utc)  # pylint: disable=maybe-no-member
-
-            return parsed_date
-
-        if value is None:
-            return None
-
-        if isinstance(value, datetime.datetime):
-            return value
+        if not isinstance(value, datetime.datetime):
+            raise TypeError(
+                "Value should be loaded from a string, a datetime object or None, not {}".format(type(value))
+            )
 
-        raise TypeError("Value should be loaded from a string, not {}".format(type(value)))
+        if value.tzinfo is not None:  # pylint: disable=maybe-no-member
+            return value.astimezone(pytz.utc)  # pylint: disable=maybe-no-member
```

Comment:
How does this interact with edx platform, where afaik there is no support
yet for timezones and everything is represented in the server's time?

On Tue, Sep 1, 2015 at 4:05 PM, John Eskew <notifications@github.com> wrote:

> In xblock/fields.py
> <https://github.com/edx/XBlock/pull/308#discussion_r38464730>:
>
> >
> > -        raise TypeError("Value should be loaded from a string, not {}".format(type(value)))
> > +        if value.tzinfo is not None:  # pylint: disable=maybe-no-member
> > +            return value.astimezone(pytz.utc)  # pylint: disable=maybe-no-member
>
> Before this change, it looks like .astimezone() was actually doing
> nothing, right? Or did parsed_date actually get its time zone changed?
> The way you've done it is definitely more correct - I just wonder about
> side-effects of code relying on the previous behavior. Perhaps you should
> add a test which sets a DateTime field to a value with a timezone different
> than UTC - does one exist now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/XBlock/pull/308/files#r38464730>.
>